File: src/device/qasm_gates.c
#include <math.h>
#include <complex.h>
#include "datatypes.h"
#include "resources.h"
#include "qasm_gates.h"
#include "quest/include/operations.h"

cq_status unitary(qubit * qh, const double THETA, const double PHI,
const double LAMBDA) {
  cq_status status = CQ_ERROR;
  qcomp z_th = cexp(I * THETA);
  qcomp a = (1.0 + z_th) / 2;
  qcomp b = -I * cexp(I*LAMBDA) * (1.0 - z_th) / 2;
  qcomp c = I * cexp(I * PHI) * (1.0 - z_th) / 2;
  qcomp d = cexp(I * (PHI+LAMBDA)) * (1.0 + z_th) / 2;
  CompMatr1 U = getInlineCompMatr1({{a, b}, {c, d}});

  if (qh != NULL) {
    applyCompMatr1(qregistry.registers[qh->registry_index], qh->offset, U);
    status = CQ_SUCCESS;
  }

  return status;
}

cq_status gphase(qubit * qh, const double THETA) {
  cq_status status = CQ_ERROR;

  qcomp z = cexp(I*THETA);
  DiagMatr1 g = getInlineDiagMatr1({z,z});

  if (qh != NULL) {
    applyDiagMatr1(qregistry.registers[qh->registry_index], qh->offset, g);
    status = CQ_SUCCESS;
  }

  return status;
}

cq_status phase(qubit * qh, const double THETA) {
  cq_status status = CQ_ERROR;

  if (qh != NULL) {
    applyPhaseShift(qregistry.registers[qh->registry_index], qh->offset, THETA);
    status = CQ_SUCCESS;
  }

  return status;
}

cq_status paulix(qubit * qh) {
  cq_status status = CQ_ERROR;

  if (qh != NULL) {
    applyPauliX(qregistry.registers[qh->registry_index], qh->offset);
    status = CQ_SUCCESS;
  }

  return status;
}

cq_status pauliy(qubit * qh) {
  cq_status status = CQ_ERROR;

  if (qh != NULL) {
    applyPauliY(qregistry.registers[qh->registry_index], qh->offset);
    status = CQ_SUCCESS;
  }

  return status;
}

cq_status pauliz(qubit * qh) {
  cq_status status = CQ_ERROR;

  if (qh != NULL) {
    applyPauliZ(qregistry.registers[qh->registry_index], qh->offset);
    status = CQ_SUCCESS;
  }

  return status;
}

cq_status hadamard(qubit * qh) {
  cq_status status = CQ_ERROR;
  
  if (qh != NULL) {
    applyHadamard(qregistry.registers[qh->registry_index], qh->offset);
    status = CQ_SUCCESS;
  }
  
  return status;
}

cq_status sqrtz(qubit * qh) {
  cq_status status = CQ_ERROR;

  if (qh != NULL) {
    applyS(qregistry.registers[qh->registry_index], qh->offset);
    status = CQ_SUCCESS;
  }

  return status;
}

cq_status sqrtzhc(qubit * qh) {
  cq_status status = CQ_ERROR;
  DiagMatr1 conjS = getInlineDiagMatr1({1, -I});

  if (qh != NULL) {
    applyDiagMatr1(qregistry.registers[qh->registry_index], qh->offset, conjS);
    status = CQ_SUCCESS;
  }

  return status;
}

cq_status sqrts(qubit * qh) {
  cq_status status = CQ_ERROR;

  if (qh != NULL) {
    applyT(qregistry.registers[qh->registry_index], qh->offset);
    status = CQ_SUCCESS;
  }

  return status;
}

cq_status sqrtshc(qubit * qh) { 
  cq_status status = CQ_ERROR;
  DiagMatr1 conjT = getInlineDiagMatr1({1, (1.0 - I)/sqrt(2)});

  if (qh != NULL) {
    applyDiagMatr1(qregistry.registers[qh->registry_index], qh->offset, conjT);
    status = CQ_SUCCESS;
  }

  return status;
}

cq_status sqrtx(qubit * qh) {
  cq_status status = CQ_ERROR;
  CompMatr1 SX = getInlineCompMatr1(
    {
      { (1 + I) / 2, (1 - I) / 2 },
      { (1 - I) / 2, (1 + I) / 2 }
    }
  );

  if (qh != NULL) {
    applyCompMatr1(qregistry.registers[qh->registry_index], qh->offset, SX);
    status = CQ_SUCCESS;
  }

  return status;
}

cq_status rotx(qubit * qh, const double THETA) {
  cq_status status = CQ_ERROR;

  if (qh != NULL) {
    applyRotateX(qregistry.registers[qh->registry_index], qh->offset, THETA);
    status = CQ_SUCCESS;
  }

  return status;
}

cq_status roty(qubit * qh, const double THETA) {
  cq_status status = CQ_ERROR;

  if (qh != NULL) {
    applyRotateY(qregistry.registers[qh->registry_index], qh->offset, THETA);
    status = CQ_SUCCESS;
  }

  return status;
}

cq_status rotz(qubit * qh, const double THETA) {
  cq_status status = CQ_ERROR;

  if (qh != NULL) {
    applyRotateZ(qregistry.registers[qh->registry_index], qh->offset, THETA);
    status = CQ_SUCCESS;
  }

  return status;
}

cq_status cpaulix(qubit * ctrl, qubit * target) {
  cq_status status = CQ_ERROR;

  if (ctrl != NULL && target != NULL && ctrl != target) {
    applyControlledPauliX(qregistry.registers[ctrl->registry_index], ctrl->offset, target->offset);
    status = CQ_SUCCESS;
  }

  return status;
}

cq_status cpauliy(qubit * ctrl, qubit * target) {
  cq_status status = CQ_ERROR;

  if (ctrl != NULL && target != NULL && ctrl != target) {
    applyControlledPauliY(qregistry.registers[ctrl->registry_index], ctrl->offset, target->offset);
    status = CQ_SUCCESS;
  }

  return status;
}

cq_status cpauliz(qubit * ctrl, qubit * target) {
  cq_status status = CQ_ERROR;

  if (ctrl != NULL && target != NULL && ctrl != target) {
    applyControlledPauliZ(qregistry.registers[ctrl->registry_index], ctrl->offset, target->offset);
    status = CQ_SUCCESS;
  }

  return status;
}

cq_status cphase(qubit * ctrl, qubit * target, const double THETA) {
  cq_status status = CQ_ERROR;

  if (ctrl != target && ctrl != NULL && target != NULL) {
    Qureg qureg = qregistry.registers[ctrl->registry_index];
    applyTwoQubitPhaseShift(qureg, ctrl->offset, target->offset, THETA);
    status = CQ_SUCCESS;
  }

  return status;
}

cq_status crotx(qubit * ctrl, qubit * target, const double THETA) {
  cq_status status = CQ_ERROR;

  if (ctrl != NULL && target != NULL && ctrl != target) {
    Qureg qureg = qregistry.registers[ctrl->registry_index];
    applyControlledRotateX(qureg, ctrl->offset, target->offset, THETA);
    status = CQ_SUCCESS;
  }

  return status;
}

cq_status croty(qubit * ctrl, qubit * target, const double THETA) {
  cq_status status = CQ_ERROR;

  if (ctrl != NULL && target != NULL && ctrl != target) {
    Qureg qureg = qregistry.registers[ctrl->registry_index];
    applyControlledRotateY(qureg, ctrl->offset, target->offset, THETA);
    status = CQ_SUCCESS;
  }

  return status;
}

cq_status crotz(qubit * ctrl, qubit * target, const double THETA) {
  cq_status status = CQ_ERROR;
  
  if (ctrl != NULL && target != NULL && ctrl != target) {
    Qureg qureg = qregistry.registers[ctrl->registry_index];
    applyControlledRotateZ(qureg, ctrl->offset, target->offset, THETA);
    status = CQ_SUCCESS;
  }
  
  return status;
}

cq_status chadamard(qubit * ctrl, qubit * target) {
  cq_status status = CQ_ERROR;
  return status;
}

cq_status cunitary(qubit * ctrl, qubit * target, const double THETA, 
const double PHI, const double LAMBDA) {
  cq_status status = CQ_ERROR;
  return status;
}

cq_status swap(qubit * a, qubit * b) {
  cq_status status = CQ_ERROR;

  if (a != b && a != NULL && b != NULL) {
    Qureg qureg = qregistry.registers[a->registry_index];
    applySwap(qureg, a->offset, b->offset);
    status = CQ_SUCCESS;
  }
  
  return status;
}

cq_status ccpaulix(qubit * ctrl_a, qubit * ctrl_b, qubit * target) {
  cq_status status = CQ_ERROR;
  return status;
}

cq_status cswap(qubit * ctrl, qubit * a, qubit * b) {
  cq_status status = CQ_ERROR;
  return status;
}

File: src/device/fortran/qasm_gates.f90
submodule (cq) qasm_gates
use c_device_interface
implicit none

contains
! --------------------------------- QASM GATES --------------------------------

  function get_qubit_at(qh, qubit_idx) result(new_ptr)
    implicit none
    type(qubit), value :: qh
    integer(c_intptr_t), value :: qubit_idx
    type(c_ptr) :: curr_ptr
    integer(c_intptr_t) :: offset
    type(c_ptr) :: new_ptr
    type(qubit_size) :: tmp

    ! validate representation of c_intptr_t and c_ptr
    integer(c_intptr_t) :: unused_i
    type(c_ptr) :: unused_p
    integer, parameter :: pointer_sized = &
    merge(c_intptr_t, -1, storage_size(unused_i) == storage_size(unused_p))
    integer(pointer_sized) :: ip
    ! end

    curr_ptr = qh%this

    offset = transfer(curr_ptr, offset) + (c_sizeof(tmp) * qubit_idx)
    new_ptr = transfer(offset, new_ptr)
  end function get_qubit_at

  module procedure cq_unitary !(qh, qubit_idx, THETA, PHI, LAMBDA) result(status)
    status = unitary(get_qubit_at(qh, qubit_idx), THETA, PHI, LAMBDA)
  end procedure cq_unitary

  module procedure cq_gphase !(qh, qubit_idx, THETA) result(status)
    status = gphase(get_qubit_at(qh, qubit_idx), THETA)
  end procedure cq_gphase

  module procedure cq_paulix !(qh, qubit_idx) result(status) 
    status = paulix(get_qubit_at(qh, qubit_idx))
  end procedure cq_paulix

  module procedure cq_pauliy !(qh, qubit_idx) result(status) 
    status = pauliy(get_qubit_at(qh, qubit_idx))
  end procedure cq_pauliy

  module procedure cq_pauliz !(qh, qubit_idx) result(status) 
    status = pauliz(get_qubit_at(qh, qubit_idx))
  end procedure cq_pauliz

  module procedure cq_hadamard !(qh, qubit_idx) result(status)
    status = hadamard(get_qubit_at(qh, qubit_idx))
  end procedure cq_hadamard

  module procedure cq_sqrtz !(qh, qubit_idx) result(status) 
    status = sqrtz(get_qubit_at(qh, qubit_idx))
  end procedure cq_sqrtz

  module procedure cq_sqrtzhc !(qh, qubit_idx) result(status) 
    status = sqrtzhc(get_qubit_at(qh, qubit_idx))
  end procedure cq_sqrtzhc

  module procedure cq_sqrts !(qh, qubit_idx) result(status) 
    status = sqrts(get_qubit_at(qh, qubit_idx))
  end procedure cq_sqrts

  module procedure cq_sqrtshc !(qh, qubit_idx) result(status) 
    status = sqrtshc(get_qubit_at(qh, qubit_idx))
  end procedure cq_sqrtshc

  module procedure cq_sqrtx !(qh, qubit_idx) result(status) 
    status = sqrtx(get_qubit_at(qh, qubit_idx))
  end procedure cq_sqrtx

  module procedure cq_rotx !(qh, qubit_idx, THETA) result(status) 
    status = rotx(get_qubit_at(qh, qubit_idx), THETA)
  end procedure cq_rotx

  module procedure cq_roty !(qh, qubit_idx, THETA) result(status) 
    status = roty(get_qubit_at(qh, qubit_idx), THETA)
  end procedure cq_roty

  module procedure cq_rotz !(qh, qubit_idx, THETA) result(status) 
    status = rotz(get_qubit_at(qh, qubit_idx), THETA)
  end procedure cq_rotz
 
  module procedure cq_cpaulix !(qr, ctrl, qtarget) result(status) 
    status = cpaulix(get_qubit_at(qr, ctrl), get_qubit_at(qr, qtarget))
  end procedure cq_cpaulix

  module procedure cq_cpauliy !(qr, ctrl, qtarget) result(status) 
    status = cpauliy(get_qubit_at(qr, ctrl), get_qubit_at(qr, qtarget))
  end procedure cq_cpauliy

  module procedure cq_cpauliz !(qr, ctrl, qtarget) result(status) 
    status = cpauliz(get_qubit_at(qr, ctrl), get_qubit_at(qr, qtarget))
  end procedure cq_cpauliz

  module procedure cq_cphase !(qr, ctrl, qtarget, THETA) result(status) 
    status = cphase(get_qubit_at(qr, ctrl), get_qubit_at(qr, qtarget), THETA)
  end procedure cq_cphase

  module procedure cq_crotx !(qr, ctrl, qtarget, THETA) result(status) 
    status = crotx(get_qubit_at(qr, ctrl), get_qubit_at(qr, qtarget), THETA)
  end procedure cq_crotx

  module procedure cq_croty !(qr, ctrl, qtarget, THETA) result(status) 
    status = croty(get_qubit_at(qr, ctrl), get_qubit_at(qr, qtarget), THETA)
  end procedure cq_croty

  module procedure cq_crotz !(qr, ctrl, qtarget, THETA) result(status) 
    status = crotz(get_qubit_at(qr, ctrl), get_qubit_at(qr, qtarget), THETA)
  end procedure cq_crotz

  module procedure cq_chadamard !(qr, ctrl, qtarget) result(status) 
    status = chadamard(get_qubit_at(qr, ctrl), get_qubit_at(qr, qtarget))
  end procedure cq_chadamard

  module procedure cq_cunitary !(qr, ctrl, qtarget, THETA, PHI, LAMBDA) result(status) 
    status = cunitary(get_qubit_at(qr, ctrl), get_qubit_at(qr, qtarget), THETA, PHI, LAMBDA)
  end procedure cq_cunitary

  module procedure cq_swap !(qr, a, b) result(status) 
    status = swap(get_qubit_at(qr, a), get_qubit_at(qr, b))
  end procedure cq_swap

  module procedure cq_ccpaulix !(qr, ctrl_a, ctrl_b, qtarget) result(status) 
    status = ccpaulix(get_qubit_at(qr, ctrl_a), get_qubit_at(qr, ctrl_b), get_qubit_at(qr, qtarget))
  end procedure cq_ccpaulix

  module procedure cq_cswap !(qr, ctrl, a, b) result(status) 
    status = cswap(get_qubit_at(qr, ctrl), get_qubit_at(qr, a), get_qubit_at(qr, b))
  end procedure cq_cswap

end submodule qasm_gates
